src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs

```diff
@@ -104,31 +101,59 @@ internal static class UseCollectionExpressionHelpers
-            var collectionBuilderType = compilation.CollectionBuilderAttribute();
-            if (structType.GetAttributes().Any(a => a.AttributeClass?.Equals(collectionBuilderType) is true))
-                return true;
+                // If it has a [CollectionBuilder] attribute on it, it is a valid collection expression type.
+                var collectionBuilderType = compilation.CollectionBuilderAttribute();
+                if (namedType.GetAttributes().Any(a => a.AttributeClass?.Equals(collectionBuilderType) is true))
+                    return true;
+
+                // At this point, all that is left are collection-initializer types.  These need to derive from
+                // System.Collections.IEnumerable, and have an accessible no-arg constructor.
+                if (namedType.AllInterfaces.Contains(compilation.IEnumerableType()!))
+                {
+                    // If they have an accessible `public C(int capacity)` constructor, the lang prefers calling that.
+                    var constructors = namedType.Constructors;
+                    var capacityConstructor = GetAccessibleInstanceConstructor(constructors, c => c.Parameters is [{ Name: "capacity", Type.SpecialType: SpecialType.System_Int32 }]);
+                    if (capacityConstructor != null)
+                        return true;
 
-            // Otherwise, presume this is unsafe to use with collection expressions.
+                    var noArgConstructor =
+                        GetAccessibleInstanceConstructor(constructors, c => c.Parameters.IsEmpty) ??
+                        GetAccessibleInstanceConstructor(constructors, c => c.Parameters.All(p => p.IsOptional || p.IsParams));
+                    if (noArgConstructor != null)
+                    {
+                        // If we have a struct, and the constructor we find is implicitly declared, don't consider this
+                        // a constructible type.  It's likely the user would just get the `default` instance of the
+                        // collection (like with ImmutableArray<T>) which would then not actually work.  If the struct
+                        // does have an explicit constructor though, that's a good sign it can actually be constructed
+                        // safely with the no-arg `new S()` call.
+                        if (!(namedType.TypeKind == TypeKind.Struct && noArgConstructor.IsImplicitlyDeclared))
+                            return true;
+                    }
+                }
+            }
+
+            // Anything else is not constructible.
             return false;
         }
+
+        IMethodSymbol? GetAccessibleInstanceConstructor(ImmutableArray<IMethodSymbol> constructors, Func<IMethodSymbol, bool> predicate)
+        {
+            var constructor = constructors.FirstOrDefault(c => !c.IsStatic && predicate(c));
+            return constructor is not null && constructor.IsAccessibleWithin(compilation.Assembly) ? constructor : null;
```

Comment:
❔ I can't find a specific location where abstract types are disallowed. In most cases, the parameterless constructor of an abstract type would not be accessible, but there are specific cases where it is (e.g. if you are currently in the implementation of a derived type, or if the constructor has been explicitly declared `public`).